src/rt/region/region_api.h

```diff
@@ -0,0 +1,332 @@
+  /**
+   * Check if pointer points to a new region.
+   */
+  inline bool is_region_ref(Object* o)
+  {
+    // Check if iso
+    if (!o->debug_is_iso())
+      return false;
+    // Check for entry point
+    return RegionContext::get_entry_point() != o;
+  }
+
+  /**
+   * Open supplied region, and return entry point.
+   */
+  inline void open_region(Object* r)
+  {
+    auto md = r->get_region();
+    RegionContext::push(r, md);
+    switch (Region::get_type(md))
+    {
+      case RegionType::Trace:
+      case RegionType::Arena:
+        break;
+      case RegionType::Rc:
+        ((RegionRc*)md)->open(r);
+        break;
+      default:
+        abort();
+    }
+  }
+
+  /**
+   * Close current region
+   */
+  inline void close_region()
+  {
+    auto md = RegionContext::get_region();
+    switch (Region::get_type(md))
+    {
+      case RegionType::Trace:
+      case RegionType::Arena:
+        break;
+      case RegionType::Rc:
+        ((RegionRc*)md)->close(RegionContext::get_entry_point());
+        break;
+      default:
+        abort();
+    }
+    RegionContext::pop();
+  }
+
+  class UsingRegion
+  {
+  public:
+    UsingRegion(Object* r)
+    {
+      open_region(r);
+    }
+
+    ~UsingRegion()
+    {
+      // TODO: Check if we are in the same region as the one we opened.
+      close_region();
+    }
+  };
+
+  /**
+   * Freeze region
+   */
+  template<typename T = Object>
+  inline T* freeze(T* r)
+  {
+    // Check for trace region.
+    Freeze::apply(ThreadAlloc::get(), r);
+    return r;
+  }
+
+  /**
+   * Add supplied region to the current region
+   * and return the entry point.
+   */
+  template<typename T = Object>
+  inline T* merge(T* r)
+  {
+    // Confirm regions are the same type
+    assert(
+      Region::get_type(r->get_region()) ==
+      Region::get_type(RegionContext::get_region()));
+
+    switch (Region::get_type(r->get_region()))
+    {
+      case RegionType::Trace:
+        RegionTrace::merge(ThreadAlloc::get(), RegionContext::get_entry_point(), r);
+        return r;
+      case RegionType::Arena:
+        RegionArena::merge(ThreadAlloc::get(), RegionContext::get_entry_point(), r);
+        return r;
+      case RegionType::Rc:
+        abort();
+    }
+  }
+
+  /**
+   * Create external reference to o in the current region.
+   */
+  inline ExternalRef* create_external_reference(Object* o)
+  {
+    return ExternalRef::create(RegionContext::get_region(), o);
+  }
+
+  /**
+   * Check if external reference is in the current region and still valid.
+   */
+  inline bool is_external_reference_valid(ExternalRef* e)
+  {
+    return e->is_in(RegionContext::get_region());
+  }
+
+  /**
+   * Use external reference e, and return the object in the current region.
+   */
+  inline Object* use_external_reference(ExternalRef* e)
+  {
+    assert(is_external_reference_valid(e));
+    return e->get();
+  }
+
+  /**
+   * Create object in current region
+   */
+  inline Object* create_object(Descriptor* d)

```

Comment:
Is this intended to replace `Region::alloc` in `region.h`? If not, why would I use one over the other?

Reply:
Should be removed.  I have not been dropping stuff unless it doesn't compile initially.